src/main/scala/org/allenai/pipeline/PipescriptParser.scala

```diff
@@ -0,0 +1,264 @@
+    def resolve(env: MMap[String, String]) = KeyValuePairsToken(keyValuePairs.resolve(env))
+  }
+
+  /** A StringExp is either:
+    * a string (quotes optional if it contains no special characters)
+    * an s-string (for substitutions within a string),
+    * or a variable reference.
+    */
+  sealed abstract class StringExp {
+    def asString: String
+
+    def resolve(environment: MMap[String, String]): LiteralString
+  }
+
+  /** i.e. ${var} */
+  case class VariableReference(name: String) extends StringExp {
+    def asString = sys.error(s"Unresolved variable $name")
+
+    def resolve(environment: MMap[String, String]): LiteralString =
+      environment.get(name).map(LiteralString.apply).getOrElse {
+        throw new IllegalArgumentException(s"Could not find variable '$name' in environment: $environment")
+      }
+  }
+
+  /** A string expressed in Java Syntax.
+    * i.e. "Hel\"o"
+    */
+  case class JavaString(s: String) extends StringExp {
+    def asString = StringHelpers.unescape(StringHelpers.stripQuotes(s))
+
+    def resolve(environment: MMap[String, String]): LiteralString = LiteralString(asString)
+  }
+
+  /** A literal string without quotes, escaping, or anything else */
+  case class LiteralString(s: String) extends StringExp {
+    def asString = s
+
+    def resolve(environment: MMap[String, String]): LiteralString = this
+  }
+
+  /** A scala-style substitution string."
+    * s"Hello ${your.name}! My name is ${myName}"
+    */
+  case class SubstitutionString(stringLiteral: String) extends StringExp {
+    val stringBody = StringHelpers.stripQuotes(stringLiteral)
+
+    def asString: String = sys.error(s"String literal with unresolved escape characters: $stringLiteral")
+
+    def resolve(env: MMap[String, String]): LiteralString = {
+      case class Replacement(regex: Regex, logic: Match => String)
+
+      def lookup(m: Match): String = {
+        val ref = m.group(1)
+        val value = env.get(ref).getOrElse {
+          throw new IllegalArgumentException(s"Could not find variable '$ref' in environment: " +
+            env)
+        }
+
+        value
+      }
+
+      val replacements = Seq(
+        Replacement(StringHelpers.bracesVariableRegex, lookup),
+        Replacement(StringHelpers.plainVariableRegex, lookup)
+      )
+
+      val replaced = replacements.foldLeft(stringBody) { (string, replacement) =>
+        replacement.regex.replaceAllIn(string, replacement.logic)
+      }
+
+      LiteralString(StringHelpers.unescape(replaced))
+    }
+  }
+
+  object StringHelpers {
+    val bracesVariableRegex = """(?<!\\)\$\{([^}]+)\}""".r
+    val plainVariableRegex = """(?<!\\)\$(\w+)""".r
+
+    def stripQuotes(s: String) = s.substring(1, s.length - 1)
+
+    def unescape(s: String) = {
+      StringEscapeUtils.unescapeJava(s)
+    }
+  }
+
+  /** The parser combinator.  This is a subclass because parser combinators are not threadsafe. */
+  class Parser extends JavaTokenParsers {
```

Comment:
Main difference is that commands can now be broken into multiple lines. End-of-line comments are also supported, but the comments are now discarded as whitespace, so we can't write them back to the portable script.